Playbooks

0x0elliot opened the pull request:
Still need to wrap up tests, linting up the code, and finish up the docs. Then leave space for code reviews and further testing.

0ssigeno commented on api_app/playbooks_manager/views.py:
```diff
@@ -0,0 +1,162 @@
+    # Check if task is test or not
+    if not settings.STAGE_CI:
+        # fire celery task
+
+        for index, job in enumerate(jobs):
+            celery_app.send_task(
+                "start_playbooks",
+                kwargs=dict(
+                    job_id=job.pk,
+                    playbooks_to_execute=cleaned_playbooks_list,
+                ),
+            )
+
+    ser = PlaybookAnalysisResponseSerializer(
+        data=[{
+            "status": "accepted",
+            "job_id": job.pk,
+            "warnings": warnings,
+            "playbooks_running": cleaned_playbooks_list,
+            "analyzers_running": analyzers_to_be_run,
+            "connectors_running": connectors_to_be_run
+        }
+            for index, job in enumerate(jobs)
+        ],
+        many=True
+    )
+    ser.is_valid(raise_exception=True)
+
+    response_dict = {
+        "count": len(ser.data),
+        "results": ser.data,
+    }
+
+    logger.debug(response_dict)
+
+    return Response(
+        response_dict,
+        status=status.HTTP_200_OK,
+    )
+
```
This is not maintainable. It is exactly the same code copy-pasted from `_multi_analysis_request` with a different output serializer and with the call to `playbooks_controller.filter_playbooks`. I'm 100% ok with having a different endpoint, but this function **must not** share its code with `_multi_analysis_request`.

0x0elliot replied:
Yes, I am looking into moving this code to one function itself. I had to rewrite a lot of this to support multiple observables and quickly get a PR open because I was seeing a lot of changes being made which made it harder to implement pre-written code.

0ssigeno commented on api_app/views.py:
```diff
@@ -27,6 +27,8 @@
 from .analyzers_manager import controller as analyzers_controller
 from .analyzers_manager.constants import ObservableTypes
+from .connectors_manager import controller as connectors_controller
+from .playbooks_manager import controller as playbooks_controller
```
Why are these here?

0x0elliot replied:
Removed.

0ssigeno commented on api_app/playbooks_manager/dataclasses.py:
```diff
@@ -0,0 +1,57 @@
+
+    @classmethod
+    def from_dict(cls, data: dict) -> "PlaybookConfig":
+        keys_to_pop = ["config", "verification", "secrets", "params", "python_module"]
+        [data.pop(key) for key in keys_to_pop]
```
I'm ok with having this utility method, but popping wrong keys makes the assumption that `data` is formatted in some way. You should, instead, follow the opposite approach: whitelist only the keys that should be parsed.

0x0elliot replied:
Way better approach! I will look into this. The code didn't require popping until the last version release. After I took care of the conflicts, This issue had to be quickly put to rest like this.

0ssigeno commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,283 @@
+logger = logging.getLogger(__name__)
+
+
+def filter_playbooks(serialized_data: Dict, warnings: List[str]) -> List[str]:
```
Wrong signature

0ssigeno commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,283 @@
+
+def filter_playbooks(serialized_data: Dict, warnings: List[str]) -> List[str]:
+    # init empty list
+    cleaned_playbook_list = []
```
I think `valid_playbook_list`, or `to_be_run_playbook_list`, or something like that are a better name for what you are trying to do

0x0elliot replied:
Alright. `valid_playbook_list` sounds better. Will use it.

0ssigeno commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,283 @@
+            if not pp:
+                if not run_all:
+                    raise NotRunnablePlaybook(
+                        f"{p_name} won't run: not available for configuration"
```
`not present in configuration` maybe?

0x0elliot replied:
My bad. That would be a better log to store.

0ssigeno commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,283 @@
+            if not pp:
+                if not run_all:
+                    raise NotRunnablePlaybook(
+                        f"{p_name} won't run: not available for configuration"
```
`not present in configuration` maybe?

0ssigeno commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,283 @@
+                continue
+            if not pp.is_ready_to_use:  # check configured/disabled
+                raise NotRunnablePlaybook(
+                    f"{p_name} won't run: is disabled or unconfigured"
```
not configured

0x0elliot replied:
Fixed

0ssigeno commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,283 @@
+logger = logging.getLogger(__name__)
+
+
+def filter_playbooks(serialized_data: Dict, warnings: List[str]) -> List[str]:
```
I really dislike having functions that changes the input parameters. In this case, you are getting `warnings` as inputs and changing it inside the function itself. I think that is better, as best practice, to not having it as input, and, returning a new warning list that the caller will extend(if it needs to)

0x0elliot replied:
Better. Getting to that.

0ssigeno commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,283 @@
+                args = [
+                    job_id,
+                    cc.asdict(),
+                    {"runtime_configuration": c_params, "task_id": task_id},
+                    p_name
+                ]
+
+                # get celery queue
+                queue = cc.config.queue
+                if queue not in settings.CELERY_QUEUES:
+                    logger.error(
+                        f"Connector {c_name} has a wrong queue."
+                        f" Setting to `{DEFAULT_QUEUE}`"
+                    )
+                queue = DEFAULT_QUEUE
+                # get soft_time_limit
+                soft_time_limit = cc.config.soft_time_limit
+                # create task signature and add to list
+                task_signatures.append(
+                    tasks.run_connector.signature(
+                        args,
+                        {},
+                        queue=queue,
+                        soft_time_limit=soft_time_limit,
+                        task_id=task_id,
+                    )
+                )
+
+            except NotRunnableConnector as e:
+                logger.warning(e)
+
+    job.update_analyzers_to_execute(analyzers_used)
+    job.update_connectors_to_execute(connectors_used)
+
+    runner = chord(task_signatures)
+    cb_signature = tasks.post_all_playbooks_finished.signature(
+        [job.pk], immutable=True
+    )
+    runner(cb_signature)
+    return None
```
Why not writing a wrapper for `start_analyzers`? I think you wrote a lot of code, that I really don't think we need to maintain. Just do your setup, call `start_analyzers` and your finally your post_execution code. Probably i'm wrong and i'm not seeing the motivation for having all this

0x0elliot replied:
I think I can add a `playbook_mode` parameter to `start_analyzers` to solve this issue because there is some extra logic that goes in there as well. Then I would be able to simply call it.

0ssigeno commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,283 @@
+    try:
+        if job.status == job.Status.FAILED:
+            raise AlreadyFailedJobException()
+
+        stats_analyzers = job.get_analyzer_reports_stats()
+        stats_connectors = job.get_connector_reports_stats()
+        stats = {}
+
+        for entry in stats_analyzers:
+            stats[entry] = (
+                stats_analyzers.get(entry, 0) + stats_connectors.get(entry, 0)
+            )
+
+        logger.info(f"[REPORT] {job.__repr__()}, status:{job.status}, reports:{stats}")
+
+        if stats["running"] > 0 or stats["pending"] > 0:
+            status_to_set = job.Status.RUNNING
+        elif stats["success"] == stats["all"]:
+            status_to_set = job.Status.REPORTED_WITHOUT_FAILS
+        elif stats["failed"] == stats["all"]:
+            status_to_set = job.Status.FAILED
+        elif stats["failed"] >= 1 or stats["killed"] >= 1:
+            status_to_set = job.Status.REPORTED_WITH_FAILS
+        elif stats["killed"] == stats["all"]:
+            status_to_set = job.Status.KILLED
+
+    except AlreadyFailedJobException:
+        logger.error(
+            f"[REPORT] {job.__repr__()}, status: failed. Do not process the report"
+        )
+
+    except Exception as e:
+        logger.exception(f"job_id: {job.pk}, Error: {e}")
+        job.append_error(str(e), save=False)
+
+    finally:
+        if not (job.status == job.Status.FAILED and job.finished_analysis_time):
+            job.finished_analysis_time = get_now()
+        job.status = status_to_set
+        job.save(update_fields=["status", "errors", "finished_analysis_time"])
```
Same code as `job_cleanup` of `analyzers_manager`. Find a way to not have all this shared code

0x0elliot replied:
It's not exactly the same. One line changes but I will still try to reduce rewriting code.

0ssigeno commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,283 @@
+    """
+
+    # get job instance
+    job = Job.objects.get(pk=job_id)
+    # execute some callbacks
+    job_cleanup(job)
```
Not calling the connectors?

0x0elliot replied:
Connectors are called automatically by the analyzers at the end of their own execution if I am not wrong?

0ssigeno commented:
Just an overall comment on the code: i really like the structure, it follows all the other IntelOwl code, and it is well structured. Nothing to change on that side, is very legit.
The most important issue that I see anyway, is shared code. There are a lot of code that you took from other parts of IntelOwl, copy-pasted it in the app, and made some slight changes here and there. The issue with that is now we have 
- double the code to maintain.
- if you make a change in the shared code, you have to remember to make the same change in the other part
As you can see, in the long run this will create a lot of issues. The best solution for this problem is to wrap the copied function and, if there is necessary, make some changes/extend this function to suit your use case. 

I will give 0 comments on the frontend, @drosetti will review that, since I have pretty much zero knowledge on that and I do not want to say useless/wrong stuff

drosetti commented:
You forgot to make the database migrations and analysis don't work. Pleas solve this problem, in this way I can run the analysis and check also the UX.

0x0elliot commented on api_app/playbooks_manager/apps.py:
```diff
@@ -0,0 +1,6 @@
+
+
+class PlaybooksManagerConfig(AppConfig):
+    default_auto_field = 'django.db.models.BigAutoField'
```
understood.

drosetti reviewed (changes requested):
You forgot to make the database migrations and analysis don't work. Pleas solve this problem, in this way I can run the analysis and check also the UX.

0x0elliot commented:
As far as I am aware, it should work by itself if you start IntelOwl in a fresh environment. Database migrations are something you would need to fix for your specific environment right? Once that's taken care of, The analysis does work.

drosetti commented:
No: when IntelOwl starts it applies the existing migrations, but it doesn't create the migrations. You can check the list of migrations in `api_app/migrations` folder. When you create you migration please give to it a name related to this feature.

0x0elliot commented:
@drosetti Hey, I have put the migrations together. Please see if this works. After I merged the code with that of the new version, I have been having some troubles with the UX working. As in, The toasts for my code aren't activating when I want them to. I would recommend that you take a look and help me with that. 

Other than that, I showed the basic look to @eshaan7 and we came to the conclusion that for now, I can proceed to make a quick PR so that at least we can move forward to working on the rest with @devmrfitz

Also, @0ssigeno or @mlodic I have mode or less made the changes requested. You should be able to make another review. Thank you for your kind feedback.

0ssigeno commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,150 @@
+
+logger = logging.getLogger(__name__)
+
+def filter_playbooks(serialized_data: Dict) -> Tuple[List[str]]:
```
Wrong signature

0x0elliot replied:
I will just leave it to be a Tuple of Lists then?

0ssigeno commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,150 @@
+                    f"{p_name} won't run: not configured"
+                )
+
+            playbook_analyzers = runnable_analyzers(pp.analyzers)
+            playbook_connectors = runnable_connectors(pp.connectors)
```
Not necessary assignment

0x0elliot replied:
I did it especially thinking that it would make it more readable. But if you suggest so, I would get to remove it.

0ssigeno replied:
I get your point, and this is trivial and just my own preference. If you think that is more readable as you written, i'm 100% ok with keeping that way

0ssigeno commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,150 @@
+    final_analyzers_used = []
+    final_connectors_used = []
+
+    playbooks = list(playbook_dataclasses.items())
+    # loop over and create task signatures
+    for p_dict in playbooks:
+        p_name = p_dict[0]
```
```suggestion
for p_name, pp in playbook_dataclasses.items()
```

0ssigeno commented on api_app/utility.py:
```diff
@@ -0,0 +1,191 @@
+from celery.canvas import Signature
+
+
+logger = logging.getLogger(__name__)
+
+def runnable_analyzers(
+    analyzers_to_execute: List[str]
+) -> List[str]:
+    analyzer_dataclass = AnalyzerConfig.all()
+    return [analyzer for analyzer in analyzers_to_execute if analyzer_dataclass.get(analyzer)]
+
```
I **really** hate having an `utilities.py`. Having this file means that every function that you do not where to put, will be put there. For example this method should be a `@classmethod` of the `AnalyzerConfig` class

0x0elliot replied:
Thank you for pointing this out. I appreciate you helping me clean the code up.

0ssigeno replied:
Dw man. I just had some bad experiences with utilities files that grows exponentially and sorting them out is a huge pain in the ass that I would gladly not doing again

0ssigeno commented on api_app/utility.py:
```diff
@@ -0,0 +1,191 @@
+    analyzer_dataclass = AnalyzerConfig.all()
+    return [analyzer for analyzer in analyzers_to_execute if analyzer_dataclass.get(analyzer)]
+
+def runnable_connectors(
+    connectors_to_execute: List[str]
+) -> List[str]:
+    connector_dataclass = ConnectorConfig.all()
+    return [
+        connector for connector in connectors_to_execute if connector_dataclass.get(connector).is_ready_to_use
+    ]
```
`ConnectorConfig` @classmethod method

0x0elliot replied:
More appropriate. Thank you.

0ssigeno commented on api_app/utility.py:
```diff
@@ -0,0 +1,191 @@
+
+
+def job_cleanup(job: Job) -> None:
+    logger.info(f"[STARTING] job_cleanup for <-- {job.__repr__()}.")
+    status_to_set = job.Status.RUNNING
+
+    try:
+        if job.status == job.Status.FAILED:
+            raise AlreadyFailedJobException()
+
+        stats = job.get_analyzer_reports_stats()
+
+        logger.info(f"[REPORT] {job.__repr__()}, status:{job.status}, reports:{stats}")
+
+        if len(job.analyzers_to_execute) == stats["all"]:
+            if stats["running"] > 0 or stats["pending"] > 0:
+                status_to_set = job.Status.RUNNING
+            elif stats["success"] == stats["all"]:
+                status_to_set = job.Status.REPORTED_WITHOUT_FAILS
+            elif stats["failed"] == stats["all"]:
+                status_to_set = job.Status.FAILED
+            elif stats["failed"] >= 1 or stats["killed"] >= 1:
+                status_to_set = job.Status.REPORTED_WITH_FAILS
+            elif stats["killed"] == stats["all"]:
+                status_to_set = job.Status.KILLED
+
+    except AlreadyFailedJobException:
+        logger.error(
+            f"[REPORT] {job.__repr__()}, status: failed. Do not process the report"
+        )
+
+    except Exception as e:
+        logger.exception(f"job_id: {job.pk}, Error: {e}")
+        job.append_error(str(e), save=False)
+
+    finally:
+        if not (job.status == job.Status.FAILED and job.finished_analysis_time):
+            job.finished_analysis_time = get_now()
+        job.status = status_to_set
+        job.save(update_fields=["status", "errors", "finished_analysis_time"])
```
Can't we make this a `Job` method?

0x0elliot replied:
Will get to that then. More appropriate.

0ssigeno commented on tests/playbooks_manager/test_views.py:
```diff
@@ -0,0 +1,16 @@
+from .. import CustomAPITestCase
+
+
+class PlaybookAppViewsTestCase(CustomAPITestCase):
+    def test_get_playbook_config(self):
+        response = self.client.get("/api/get_playbook_configs")
+        self.assertEqual(response.status_code, 200)
+        self.assertNotEqual(response.json(), {})
+        self.assertDictEqual(
+            response.json(), PlaybookConfigSerializer.read_and_verify_config()
```
+12k lines written and a single test?

0x0elliot replied:
Good reminder. Let me wrap up the test cases.

0ssigeno commented on api_app/utility.py:
```diff
@@ -0,0 +1,191 @@
+    # loop over and create task signatures
+    for c_name, cc in connector_dataclasses.items():
+        # if disabled or unconfigured (this check is bypassed in STAGE_CI)
+        if not settings.STAGE_CI:
+            continue
+
+        # get runtime_configuration if any specified for this analyzer
+        runtime_params = runtime_configuration.get(c_name, {})
+        # gen a new task_id
+        task_id = uuid()
+        # construct args
+        args = [
+            job_id,
+            cc.asdict(),
+            {"runtime_configuration": runtime_params, "task_id": task_id},
+        ]
+        # get celery queue
+        queue = cc.config.queue
+        if queue not in settings.CELERY_QUEUES:
+            logger.error(
+                f"Connector {c_name} has a wrong queue."
+                f" Setting to `{DEFAULT_QUEUE}`"
+            )
+            queue = DEFAULT_QUEUE
+        # get soft_time_limit
+        soft_time_limit = cc.config.soft_time_limit
+        # create task signature and add to list
+        task_signatures.append(
+            tasks.run_connector.signature(
+                args,
+                {},
+                queue=queue,
+                soft_time_limit=soft_time_limit,
+                task_id=task_id,
+                ignore_result=True,  # since we are using group and not chord
+            )
+        )
+        connectors_used.append(c_name)
+    
+    return task_signatures, connectors_used
```
I'm thinking how we can remove this lines from there, like i said before I really hate having a "dump" file where we put functions that we do not where to put. This is prone to having a "god utilities" where the file size will increase exponentially. Once I have a solution, i will comment this

0x0elliot replied:
Okay sure. I will ask for more opinions from matteo then.

0x0elliot replied:
if the file would have grown, it would have been a hard file to get back to with a lot of junk. i appreciate you pointing out how it could be further cleaned up.

0ssigeno replied:
Unfortunately I can see the decision on why these methods should be on a common file between the two callers. I'm taking a bit of time to decide on how to structure this, I do not like to give suggestion and then understand that the suggestion itself was shit

0x0elliot commented:
Thank you for the code review. It's a nice learning experience to have your code criticized to make it as readable as possible by someone as experienced as you folks!

drosetti reviewed (changes requested):
@0x0elliot here is my review, excuse me if I took lots of time.

Talking about the UX I think it can be improved changing the UI part where users start a palybooks analysis, but if you and @eshaan7 agreed that is enough for now and proceed with other tasks, I don't want to change your plans.

However, there are two bugs to fix:
* The first bug is the one you reported: the analysis start, but the toasts aren't shown, this happen because there is an error in the function `createPlaybookJob` when you handle the response.
* The second bug is in "Plugin" section: the icons in the "Info" column shows correctly only the "Playbooks" cards, but with "Analyzers" and "Connectors" and error is raised.

mlodic commented:
please pull recent changes from `develop` because there are still incostintencies. We won't merge anything else before this PR so  I assure you this is the last time you need to pull from there.

One example is that there are incosistent migrations:

> intelowl_uwsgi           | CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0010_custom_config, 0010_playbooks in api_app).

You can just put 0010_playbooks in api_app after 0010_custom_config

mlodic commented:
Also please populate the PR template cause there are other things to be handled like the docs.

(Please add to the PR name the issue/s that this PR would close if merged by using a [Github](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) keyword. Example: `closes #999`. If your PR is made by a single commit, please add that clause in the commit too. This is all required to automate the closure of related issues.)

# Description

Please include a summary of the change.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [ ] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [ ] The pull request is for the branch `develop`
- [ ] A new analyzer or connector was added, in which case:
    - [ ] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](./Advanced-Usage.md) was updated (in case the analyzer/connector provides additional optional configuration).
    - [ ] Secrets were added in [env_file_app_template](https://github.com/intelowlproject/IntelOwl/blob/master/docker/env_file_app_template), [env_file_app_ci](https://github.com/certego/IntelOwl/blob/master/docker/env_file_app_ci) and in the [Installation](./Installation.md) docs, if necessary.
    - [ ] If the analyzer/connector requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [ ] If a File analyzer was added, its name was explicitly defined in [test_file_scripts.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_file_scripts.py) (not required for Observable Analyzers).
- [ ] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [ ] The tests gave 0 errors.
- [ ] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](./Contribute.md)).

### Important Rules
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review

# Real World Example

Please delete if the PR is for bug fixing.
Otherwise, please provide the resulting raw JSON of a finished analysis (and, if you like, a screenshot of the results). This is to allow the maintainers to understand how the analyzer works.

mlodic commented on tests/playbooks_manager/test_views.py:
```diff
@@ -0,0 +1,35 @@
+
+    # There is still space for file playbooks left here.
+    # I will be implementing them after we have a standard
+    # file analyzer to ourselves.
```
yeah the project seems incomplete without proper testing. While we still need to initialize frontend tests and that's already a big issue in IntelOwl, at least we need to properly test new things in the backend because the framework does provide everything to do that easily.

You could just take the most important functions you implemented in all the code and write tests that verify that the use cases that you wants to handle are correctly handled:
1 - hypothesis
2 - initialize required data/parameters
3 - call function which needs testing
4 - verifiy results

In tests you do not have to care about code replication. For instance, if a specific function would give you 5 different results based on the different parameters that you provide, you would need to do 5 different tests for that, with the code really similar between tests. Each test should live its own life and be autonomous from the other.

For instance, the `job_cleanup` you implemented handle a lot of different possible outcomes. You need to verify those with tests.

mlodic replied:
also `stack_connectors` and `stack_analyzers` are good targets

mlodic replied:
the `before_run` too, you changed it. You should verify all the possible outcomes. Etc

0x0elliot replied:
hmm. I am thinking of writing another test case that uploads files first. I did verify stack_connectors and stack_analyzers manually. And before_run was functionally verified after I compiled together the free analyzer playbook. And hopefully, just running the free analyzer playbook should do because it has all possible outcomes since it runs all possible analyzers.

mlodic replied:
> hmm. I am thinking of writing another test case that uploads files first

ok!

> I did verify stack_connectors and stack_analyzers manually.

Don't get me wrong. I am not saying that you have not verified this manually. I am confident about that.
But that's not the same thing of writing unit tests. Unit tests are repeatable and allow to verify that, in the future, with new changes, what you have done now won't work differently. It also allow to avoid to break things while doing new features. They are as much important as the code itself that you have written.

> And before_run was functionally verified after I compiled together the free analyzer playbook. And hopefully, just running the free analyzer playbook should do because it has all possible outcomes since it runs all possible analyzers.

```
        if self._config.type == "file":
            if not job.file:
                raise UnsupportedObservableException(
                    f"No file was given for {self._config.name} file analyzer"
                )

            elif job.file_mimetype not in self._config.supported_filetypes:
                self.report.update_status(status=self.report.Status.FAILED)
                raise UnsupportedObservableException(
                    f"Observable {self._job.file_name} is not supported"
                )

            else:
                logger.info(
                    f"{job.file_name} is supported by the analyzer {self._config.name}"
                )
                return

        elif job.observable_classification not in self._config.observable_supported:
            self.report.update_status(status=self.report.Status.FAILED)
            raise UnsupportedObservableException(
                f"Observable {job.observable_name} is not supported"
            )
```

these are possible outcomes. Tests should verify that if those conditions apply, also if the exception raise (`self.assertException`). What you are saying is that the `before_run` code is actually run by your test. That's good for coverage but we are not doing "unit tests" there. Unit tests should pick different function, independently from the rest of the code, and test them with all the possible parameters, and see if the outcomes are what are expected

mlodic reviewed (changes requested):
good job man! come on, we are about to finish :) last efforts

0x0elliot commented on frontend/yarn.lock:
```diff
@@ -0,0 +1,10719 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Noted.

0x0elliot replied:
I would just add this file in the .gitignore file for the frontend.

0x0elliot replied:
and then untrack it

0x0elliot commented on docs/source/Usage.md:
```diff
@@ -339,3 +339,7 @@ The following is the list of the available connectors. You can also navigate the
+
+### Available Playbooks
+
+Playbooks are designed to be easy to share sequence of running Analyzers/Connectors on a particular kind of observable.
```
Will do both this and the code doctor additions.

mlodic replied:
a link here to the other section that you wrote "how to add a new playbook" would help the user to find all that they need

mlodic replied:
@0x0elliot could you integrate this please? should be fast

mlodic commented:
Without any error that broke the build, I finally managed to review the GUI and the user experience. This PR has been broken for a about a month and we wasted a lot of time without doing any progress. It is time to make this PR right so we can finally merge it.

Unfortunately, I am sad to say that the overall user experience imho is really bad. We cluttered the "Scan" page with unnecessary content and we never thought about optimizing it.
Adding the following lines is not good at all:
![image](https://user-images.githubusercontent.com/30625432/188649312-ba03949d-d660-4ee9-bc8e-b8e8ccc7a94e.png)

The result for the user is complete chaos and the overall result is worse than before. Better not having this feature than having it put in the GUI in this way.
So, now we need to find a right way optimize this page and avoid to make it impossible to use.
First: the "Launch playbooks" button should not exist. The "Start Scan" button should do the job. "But there are 2 different APIs, one for the playbooks and one for the normal scan". We need to handle this. If the user selected at least a playbook, the playbooks would be run. Otherwise the analyzers/connectors. Plus, it makes no sense to set as default that all the playbooks are triggered. Playbooks are a combination of analyzers/connectors that could overlap over each other so basically it makes absolutely no sense. The user should explicitly select them.

Then, how to avoid to have all of this together? 
![image](https://user-images.githubusercontent.com/30625432/188650860-76297aaf-3826-4ec5-9eec-56a8eace9346.png)
Analyzers and connectors together make sense but Playbooks does not. Either you select playbooks or you select analyzers/connectors so logically we should split the view.
The very basic idea that I have is to do something similar to what we already do for the observables/files radio button.
I would add a new radio button with 2 options: 1-analyzers/connectors 2-playbooks
In the first case, the user would see only the selectors of analyzers and connectors. In the second case it would see only the playbook selector.

I'll edit this as soon as I complete my review

mlodic commented on frontend/src/components/jobs/result/utils/JobOverview.jsx:
```diff
@@ -15,14 +15,14 @@ export default function JobOverview({ isRunningJob, job, refetch }) {
         <strong>Connectors Report</strong>
         <Badge className="ms-2">
           {job.connectors_to_execute?.length} /&nbsp;
-          {job.connectors_requested?.length || "all"}
```
the result of this would be:
* now you have `0` instead of `all` when running playbooks. So this is still wrong
* and more now if you run all the analyzers, you would have `0` instead of `all`.
So the overall result of this change is that is worse than before.

I think that you should get here when you are running playbooks. In that case, manage this differently: the `connectors_requested` and `analyzers_requested` would not exist in that case. I think you could simply use `connectors_to_execute` and `analyzers_to_execute` instead.

0x0elliot replied:
Works? 
![image](https://user-images.githubusercontent.com/60684641/189109389-e9886384-9920-4e0a-bef9-c903b89561f0.png)

0x0elliot replied:
let me know if you want anything different when no connectors are added to the playbook.

0x0elliot replied:
Would also take care of this portion the same way (plugins executed)
![image](https://user-images.githubusercontent.com/60684641/189111527-05b959b2-847b-45d2-8aa5-1db9ca844315.png)

mlodic replied:
> Works?
that's good

0x0elliot replied:
Seems alright to me now. 
![image](https://user-images.githubusercontent.com/60684641/189174217-8cfa628b-bfb4-4142-a46f-060629fa890c.png)

mlodic commented on frontend/package.json:
```diff
@@ -2,7 +2,7 @@
   "version": "0.1.0",
   "private": true,
-  "proxy": "http://localhost:80/",
+  "proxy": "http://20.163.101.67",
```
care about that :P

mlodic replied:
ok you noticed :P

0x0elliot commented:
Progress on the UI
![image](https://user-images.githubusercontent.com/60684641/189543172-b97c9fed-d693-4834-a4b0-f0d13d7b0976.png)

0x0elliot commented:
Slight update
![image](https://user-images.githubusercontent.com/60684641/189543584-f1429ffb-ec38-4c5e-a7c7-3bd7370743a8.png)

0x0elliot commented:
quick question, do i just let this warning stay?

`>   Line 155:67:  '_quota' is assigned a value but never used  no-unused-vars`

https://github.com/intelowlproject/IntelOwl/blob/ba8a4f42bc60a3e5b6b6c286a540f38e6e6b9bfc/frontend/src/components/scan/ScanForm.jsx#L140 never really calls `_quota` again in this code.

0x0elliot commented:
hah! emojis in commit messages sure do look nice :)

0x0elliot commented:
i fixed all known frontend bugs from my side. fixed the UX as well. what's left is the backend now!! (and mentor reviews of course. there is always something new to fix there.) :tada: :tada: :tada: :tada:

0x0elliot commented:
More todos:

- [ ] modify test cases that support run_all for playbooks disabled
- [ ] fix failing test cases for backend-tests after figuring out why

mlodic commented:
hey there! new changes mmmm looking into them :P

> Progress on the UI
Great, so you separated playbooks and analyzers/connectors.
Some little tweak to consider:

- [ ] `s` is missing in analyzers/connectors
- [ ] why "select playbooks" input field is not aligned with the others?
- [ ] I would still use the same "Scan" button for both visualizations (analyzers and playbooks). It makes no sense for an user to have 2 different buttons there because they both start an analysis. Plus, in the playbook case it should be disabled if no playbooks are selected.
- [ ] is there a reason why TLP is removed for playbook? Because it shouldn't. All the logic outside analyzer/connectors selection should be exactly the same between playbooks and classic analysis. Playbooks should only allow fast selection of analyzers/connectors, nothing else. All the rest should be the same

0x0elliot commented:
noted. will implement.

0x0elliot commented:
works? 
![image](https://user-images.githubusercontent.com/60684641/190013917-d9a242f7-5e52-44e5-95a9-489eb98c5d99.png)

0x0elliot commented:
neat. time to work on the backend. feel free to review for improvements @mlodic

mlodic commented:
checking details on the frontend. Overall result: cool!

"Scan page":

- [ ] I think my explanation wasn't the most comprehensible so I have a screenshot of how was my idea about the separation of the sections there
![image](https://user-images.githubusercontent.com/30625432/190097815-3cde7195-27a8-4589-bbdd-b2076c4deaa5.png)
As you can see here I logically separated the 2 sections (observables and analyzers/connectors) that change if I change my selection in the radio buttons. In that way it is easier for the user to get what changes between visualizations. That's a fast change to do in the code I guess.
- [ ] Other question: I see that for playbooks the "Extra configuration" section was removed. This is because we are not doing a check for already existing playbooks running, right? I am talking about the `ask_multi_analysis_availability` function that we call before running analyzers/connectors normaly. Logically, it would make sense to have the same check for playbooks. But I understand that this would add additional burden. Let's ignore this for now. I'll open an issue
- [ ] Tags selection in playbooks do not work but this is caused to the problem we already talked about (we need to use the same validation/serializers that we use for classic analysis)

0x0elliot commented:
> I think my explanation wasn't the most comprehensible so I have a screenshot of how was my idea about the separation of the sections there

got it. i will just shift my code to that part. 

>  Other question: I see that for playbooks the "Extra configuration" section was removed. This is because we are not doing a check for already existing playbooks running, right? I am talking about the ask_multi_analysis_availability function that we call before running analyzers/connectors normaly. Logically, it would make sense to have the same check for playbooks. But I understand that this would add an additional burden. Let's ignore this for now. I'll open an issue

yes doing it right now wouldn't be right. but speaking of this, i wanted to bring this up to your attention but i think we should make it so that instead of it using cached results for a whole playbook, this feature takes the cached result for individual analyzers. this is a better solution now that we are moving in the direction of bulk searching. this is of course for whenever it is done.

>  Tags selection in playbooks do not work but this is caused to the problem we already talked about (we need to use the same validation/serializers that we use for classic analysis)

yes of course. this will be fixed with the backend!

0x0elliot commented:
How the new frontend looks like
![image](https://user-images.githubusercontent.com/60684641/190961325-385aeaba-6b2c-4bd7-8349-55a7e2d4e084.png)

0x0elliot commented:
the bug should be fixed. next issue remains test cases here.

0x0elliot commented:
Found 2 bugs. would solve them immediately. One is in file scan and another is simply the job starting without playbooks being provided.

0x0elliot commented:
Fixed. Now, time to fix these goddamn test cases I broke.

0x0elliot commented on api_app/core/models.py:
```diff
@@ -30,8 +30,12 @@ class AbstractReport(models.Model):
     task_id = models.UUIDField()  # tracks celery task id
 
+    # If a playbook ran the process
+    parent_playbook = models.CharField(max_length=128, null=True)
```
nope, this change broke my code.

0x0elliot commented:
No idea why I am not running the test cases properly to begin fixing them
![image](https://user-images.githubusercontent.com/60684641/193166614-13df0819-b10c-4322-bc56-cd4611ab91df.png)

mlodic commented on api_app/serializers.py:
```diff
@@ -176,16 +179,20 @@ def filter_connectors(self, serialized_data: Dict) -> List[str]:
 
-        # get values from serializer
         connectors_requested = serialized_data.get("connectors_requested", [])
+        run_all = len(connectors_requested) == 0
```
could you add a comment regarding why the logic regardin `run_all` is different between analyzers and connectors? I know this is right but for future developers this would be really hard to understand.

0x0elliot replied:
it's the same as analyzers. i will make it clear in both `filter_analyzers` and `filter_connectors` functions then

mlodic commented on api_app/views.py:
```diff
@@ -51,12 +56,21 @@ def _multi_analysis_request(
+        Type[ObservableAnalysisSerializer],
+        Type[FileAnalysisSerializer],
+        Type[PlaybookObservableAnalysisSerializer],
+        Type[PlaybookFileAnalysisSerializer],
```
cool thing this

mlodic commented on api_app/views.py:
```diff
@@ -106,28 +121,55 @@ def _multi_analysis_request(
+                    "start_playbooks",
+                    kwargs=dict(
+                        job_id=job.pk,
+                    ),
```
`runtime_configuration` is still not handled here.

mlodic commented on api_app/views.py:
```diff
@@ -106,28 +121,55 @@ def _multi_analysis_request(
+                )
+
+    if playbook_scan:
+        ser = PlaybookAnalysisResponseSerializer(
```
inheritance is cool here

mlodic commented on api_app/views.py:
```diff
@@ -106,28 +121,55 @@ def _multi_analysis_request(
+                {
+                    "status": "accepted",
+                    "job_id": job.pk,
+                    "warnings": warnings,
```
this list (warnings) is empty. we should avoid to rewrite this code.

Just declare:
```
                {
                    "status": "accepted",
                    "job_id": job.pk,
                    "warnings": serialized_data[index]["warnings"],
                    "analyzers_running": job.analyzers_to_execute,
                    "connectors_running": job.connectors_to_execute,
                }
```
before the if/else and then add the `playbooks_running` field only for playbooks

0x0elliot replied:
on it

mlodic commented on docs/source/Contribute.md:
```diff
@@ -268,6 +268,35 @@ Please see [Connectors customization section](./Usage.md#connectors-customizatio
 
+## How to add a new Playbook
+
+You may want to look at the existing [playbook_configuration.json](https://github.com/intelowlproject/IntelOwl/blob/master/configuration/playbook_config.json) file. To set up a new Playbook, Add a new entry to the configuraions file in alphabetical order:
```
typo in configuration

mlodic commented on docs/source/Tests.md:
```diff
@@ -0,0 +1,86 @@
+# Tests
```
this file is a bad merge from upstream. Please remove this

mlodic commented:
did the test run locally?

0x0elliot commented on api_app/analyzers_manager/dataclasses.py:
```diff
@@ -104,3 +112,81 @@ def all(cls) -> typing.Dict[str, "AnalyzerConfig"]:
+                job_id,
+                config.asdict(),
+                {"runtime_configuration": runtime_params, "task_id": task_id},
+                parent_playbook,
```
i don't get this. it is being treated like any other parameter. do you mean, not like any other explicit parameter and instead with the help of kwargs?

mlodic replied:
yes

```
{"runtime_configuration": runtime_params, "task_id": task_id, "parent_playbook":parent_playbook}
```

0x0elliot commented on api_app/core/classes.py:
```diff
@@ -112,14 +112,15 @@ def get_exceptions_to_catch(self) -> list:
         """
         try:
             self.before_run()
+            self.report.parent_playbook = parent_playbook
```
noted and adding

0x0elliot commented:
Nope. tests aren't running as expected locally when i am using the "test" option through `start.py`.

somehow, it keeps on running failing at `test_start_analyzers` but giving a different error than GitHub actions show.

command used: `coverage run manage.py test tests.analyzers_manager.test_observable_scripts`

```
[REPORT] (poll #0) 
>>> Job:1, status:'running' 
>>> analyzer_reports:{'all': 0, 'failed': 0, 'pending': 0, 'running': 0, 'success': 0, 'killed': 0} 
>>> connector_reports:{'all': 0, 'failed': 0, 'pending': 0, 'running': 0, 'success': 0, 'killed': 0}  
>>> Running/Pending analyzers: [] 
>>> Running/Pending connectors: []
[REPORT] (poll #1) 
>>> Job:1, status:'running' 
>>> analyzer_reports:{'all': 0, 'failed': 0, 'pending': 0, 'running': 0, 'success': 0, 'killed': 0} 
>>> connector_reports:{'all': 0, 'failed': 0, 'pending': 0, 'running': 0, 'success': 0, 'killed': 0}  
>>> Running/Pending analyzers: [] 
>>> Running/Pending connectors: []
[REPORT] (poll #2) 
>>> Job:1, status:'running' 
>>> analyzer_reports:{'all': 0, 'failed': 0, 'pending': 0, 'running': 0, 'success': 0, 'killed': 0} 
>>> connector_reports:{'all': 0, 'failed': 0, 'pending': 0, 'running': 0, 'success': 0, 'killed': 0}  
>>> Running/Pending analyzers: [] 
>>> Running/Pending connectors: []

```

0x0elliot commented:
looks like the connectors are breaking. let me get into that.

0x0elliot commented:
this should make it work.

0x0elliot commented:
quick things i have to point out. why is `read_and_verify_config` forever cached? i.e it's timeout is set to `sys.maxsize`. for me, that's `9223372036854775807`

0x0elliot commented:
fixed all that finally. will get done with playbook test cases and add them to workflows.

mlodic commented:
> Nope. tests aren't running as expected locally when i am using the "test" option through `start.py`.
> 
> somehow, it keeps on running failing at `test_start_analyzers` but giving a different output, but the same error as GitHub actions show.
> 
> command used: `coverage run manage.py test tests.analyzers_manager.test_observable_scripts`
> 
> ```
> [REPORT] (poll #0) 
> >>> Job:1, status:'running' 
> >>> analyzer_reports:{'all': 0, 'failed': 0, 'pending': 0, 'running': 0, 'success': 0, 'killed': 0} 
> >>> connector_reports:{'all': 0, 'failed': 0, 'pending': 0, 'running': 0, 'success': 0, 'killed': 0}  
> >>> Running/Pending analyzers: [] 
> >>> Running/Pending connectors: []
> [REPORT] (poll #1) 
> >>> Job:1, status:'running' 
> >>> analyzer_reports:{'all': 0, 'failed': 0, 'pending': 0, 'running': 0, 'success': 0, 'killed': 0} 
> >>> connector_reports:{'all': 0, 'failed': 0, 'pending': 0, 'running': 0, 'success': 0, 'killed': 0}  
> >>> Running/Pending analyzers: [] 
> >>> Running/Pending connectors: []
> [REPORT] (poll #2) 
> >>> Job:1, status:'running' 
> >>> analyzer_reports:{'all': 0, 'failed': 0, 'pending': 0, 'running': 0, 'success': 0, 'killed': 0} 
> >>> connector_reports:{'all': 0, 'failed': 0, 'pending': 0, 'running': 0, 'success': 0, 'killed': 0}  
> >>> Running/Pending analyzers: [] 
> >>> Running/Pending connectors: []
> ```

I see that you managed to solve the problem. GJ! Out of curiosity, what was the cause?

mlodic commented:
> quick things i have to point out. why is `read_and_verify_config` forever cached? i.e it's timeout is set to `sys.maxsize`. for me, that's `9223372036854775807`. we should add an option to disable it while testing it. or set it very low when in testing mode.

you could change it in 
```
timeout=settings.CONFIG_CACHE_TIME if settings.DEBUG else sys.maxsize
```
and then allow CONFIG_CACHE_TIME to be populated at start

mlodic commented on api_app/playbooks_manager/controller.py:
```diff
@@ -0,0 +1,129 @@
+logger = logging.getLogger(__name__)
+
+
+def start_playbooks(
```
coverage report states that this function is not tested at all. Could you add a test for this?

0x0elliot replied:
let me write a general one for this.

0x0elliot commented:
> I see that you managed to solve the problem. GJ! Out of curiosity, what was the cause?

i don't remember the root cause for this one, but for the tests failing in general for the CI, it was pretty much an AND -> OR change somewhere in the code to allow connectors to be picked up when CI is running. apparently, i broke it somewhere along the way. then i proceeded to add some more changes to fix some stuff up.

0x0elliot commented:
encountered weird errors for ClamAV. you might need to test that one. couldn't myself because the container keeps dying on my current machine. 

the error is on this line, explaining how the reports variable is a dictionary that can't be split. then why is it being split in the first place? ClamAV isn't covered in the file analyzers test cases as well.

https://github.com/intelowlproject/IntelOwl/blob/ef40427495d6fb59ce7a3fbba6fd2e948e3e0936/api_app/analyzers_manager/file_analyzers/clamav.py#L29

0x0elliot commented:
ready for a merge?

mlodic commented:
I have been busy again, I'll check this tomorrow afternoon

mlodic commented:
- [ ] can you please fix the lines like I pushed here? should be a fast change

![image](https://user-images.githubusercontent.com/30625432/194324045-84ab1f0b-454d-4c37-901a-d1a5e8ea25b8.png)

mlodic commented on api_app/analyzers_manager/migrations/0003_alter_analyzerreport_parent_playbook.py:
```diff
@@ -0,0 +1,18 @@
+            name='parent_playbook',
+            field=models.CharField(null=True, max_length=128),
+        ),
+    ]
```
@0x0elliot these 2 are still not squashed

mlodic commented on api_app/connectors_manager/migrations/0003_alter_connectorreport_parent_playbook.py:
```diff
@@ -0,0 +1,18 @@
+            name='parent_playbook',
+            field=models.CharField(null=True, max_length=128),
+        ),
+    ]
```
@0x0elliot these 2 are still not squashed

mlodic commented on tests/playbooks_manager/test_controller.py:
```diff
@@ -0,0 +1,55 @@
+    test_job : Job
+
+    def setUp(self):
+        playbooks_to_test = os.environ.get("TEST_PLAYBOOKS", "").split(",")
```
please add this custom value in the env_file_app_template and in the Contribute.md with the other

0x0elliot replied:
oop. how did i miss pushing that. i did add it to the ci file

mlodic commented on frontend/package.json:
```diff
@@ -2,7 +2,7 @@
   "version": "0.1.0",
   "private": true,
-  "proxy": "http://localhost:80/",
+  "proxy": "http://hi-elliot.westus3.cloudapp.azure.com/",
```
oh noticed this again :P

0x0elliot replied:
oh my :P

0x0elliot commented:
please review once again.

mlodic commented:
please squash migrations and I'll merge! OMG I cant't believe it :) I am so excited to close this!

mlodic commented on .github/ISSUE_TEMPLATE/new_analyzer.md:
```diff
@@ -14,6 +14,8 @@ assignees: ''
 ## Type of analyzer
 **this can be observable, file, and docker**
 
+**Please Note:** If the analyzer is free, Please add it in the `FREE_TO_USE_ANALYZERS` playbook in `playbook_config.json`
```
ah one last thing, this should be put here: https://github.com/intelowlproject/IntelOwl/blob/master/.github/release_template.md. Can you move that please?

0x0elliot replied:
yessir let me do that quickly

0x0elliot commented:
🎉 🎉 🎉 🎉 🎉  and we are done with Playbooks now!

mlodic commented:
squashing was not done properly, I'll adjust it and merge
